core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1780,6 +1782,8 @@ class KafkaApis(val requestChannel: RequestChannel,
         }
 
         val controlRecords = mutable.Map.empty[TopicIdPartition, MemoryRecords]
+        val markerTransactionVersion = marker.transactionVersion()
+
```

Comment:
We should just do `val producerEpoch = if (marker.transactionVersion < 2) marker.producerEpoch else marker.producerEpoch - 1` and add a comment that for TV2+ we expect proper transaction markers to have bumped epoch, so we need to undo the bump so that comparison can fence transaction markers from previous transaction that may arrive with current epoch.

Then most of the other changes can be removed.

This would also work for `groupCoordinator.completeTransaction` code path.

Reply:
You're right it wouldn't work, at this level we're checking and updating the same epoch.  But we do need to address groupCoordinator.completeTransaction code path.